### PR TITLE
notification: improved API

### PR DESCRIPTION
**Pull request by waht**
# Purpose

This PR improves the notification API. It adds new types (both C and kdbtypes.h). It also adds macros for generating API functions for new types and also adds C99 definitions to kdbtypes.h.

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I added unit tests
- [x] I ran all tests and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] release notes are updated (doc/news/_preparation_next_release.md)
- [x] rebase when dbus PR (#1825) is merged into `master`.
- [x] update notification tutorial: explain emergent misbehavior & add guidelines
- [x] update architecture diagram

@markus2330 please review my pull request

### CODE REVIEW

**Comment by waht** [comment]
jenkins build multiconfig-gcc-stable please

**Comment by waht** [comment]
jenkins build multiconfig-gcc47-cmake-options please

**Comment by waht** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Contains commits from #1825 please rebase. I could not do a review because the shown diff is huge.

Please also fix the leftovers of #1825 here.

**Comment by waht** [comment]
Okay, I've fixed the leftovers and rebased the PR - the diff is much smaller now :)

File: src/include/kdbtypes.h [hunk 1]
```diff
@@ -48,8 +48,44 @@ using octet_t = uint8_t; // default: 0
...
+typedef uint8_t kdb_octet_t;
+typedef bool kdb_boolean_t;
+typedef int16_t kdb_short_t;
+typedef int32_t kdb_long_t;
+typedef int64_t kdb_long_long_t;
+typedef uint16_t kdb_unsigned_short_t;
+typedef uint32_t kdb_unsigned_long_t;
+typedef uint64_t kdb_unsigned_long_long_t;
+
+#if SIZEOF_LONG == 4
```
Comment (markus2330) [inline src/include/kdbtypes.h]: The `#if` is not necessary. In C99 inttypes.h defines macros for the printf modifiers, e.g., PRIu8

File: src/plugins/internalnotification/internalnotification.c [hunk 2]
```diff
@@ -153,6 +192,31 @@ static KeyRegistration * elektraInternalnotificationAddNewRegistration (PluginSt
...
+{
+	int result = 0;
+	Key * current;
+	ksRewind (ks);
+	while ((current = ksNext (ks)) != NULL)
+	{
+		if (checkKeyIsBelowOrSame (check, current))
+		{
+			result = 1;
+			break;
```
Comment (markus2330) [inline src/plugins/internalnotification/internalnotification.c]: `return 1;`

File: src/plugins/internalnotification/internalnotification.c [hunk 3]
```diff
@@ -153,6 +192,31 @@ static KeyRegistration * elektraInternalnotificationAddNewRegistration (PluginSt
...
+		}
+	}
+	return result;
```
Comment (markus2330) [inline src/plugins/internalnotification/internalnotification.c]: `return 0;`

File: src/plugins/internalnotification/internalnotification.c [hunk 4]
```diff
@@ -170,167 +234,136 @@ void elektraInternalnotificationUpdateRegisteredKeys (Plugin * plugin, KeySet *
...
 		}
 		else
 		{
-			const char * currentValue = keyString (key);
-			changed = registeredKey->lastValue == NULL || strcmp (currentValue, registeredKey->lastValue) != 0;
+			key = ksLookupByName (keySet, registeredKey->name, 0);
+			if (key == NULL)
+			{
+				registeredKey = registeredKey->next;
+				continue;
```
Comment (markus2330) [inline src/plugins/internalnotification/internalnotification.c]: some refactoring here?

File: src/plugins/internalnotification/internalnotification.c [hunk 5]
```diff
@@ -170,167 +234,136 @@ void elektraInternalnotificationUpdateRegisteredKeys (Plugin * plugin, KeySet *
...
+			ElektraNotificationChangeCallback callback = *(ElektraNotificationChangeCallback) registeredKey->callback;
+			callback (key, registeredKey->context);
 		}
 
 		registeredKey = registeredKey->next;
 	}
 }
 
+// Generate register and conversion functions
+INTERNALNOTIFICATION_TYPE (int, long int, Int, (strtol (string, &end, 10)),
```
Comment (markus2330) [inline src/plugins/internalnotification/internalnotification.c]: Very compact, well done!

Is this somehow reusable for other places? (maybe as [super macro](http://s11n.net/papers/supermacros_cpp.pdf)?)

File: src/plugins/internalnotification/testmod_internalnotification.c [hunk 6]
```diff
@@ -382,6 +629,37 @@ int main (int argc, char ** argv)
...
+	RUN_TYPE_TESTS (KdbLongDouble)
+
+	printf ("\nelektraInternalnotificationDoUpdate\n-----------------------------------\n");
+	test_doUpdateShouldUpdateKey ();
+	test_doUpdateShouldUpdateKeyBelow ();
+	test_doUpdateShouldNotUpdateKeyAbove ();
+	test_doUpdateShouldNotUpdateUnregisteredKey ();
+	test_doUpdateShouldUpdateKeyAbove ();
+
 	printf ("\ntestmod_internalnotification RESULTS: %d test(s) done. %d error(s).\n", nbTest, nbError);
```
Comment (markus2330) [inline src/plugins/internalnotification/testmod_internalnotification.c]: should use the print from the test suite.

File: src/include/kdbtypes.h (hunk 1 above)
Comment (waht) [reply src/include/kdbtypes.h]: Great, I didn't know about these!

File: src/include/kdbnotificationinternal.h [hunk 7]
```diff
@@ -23,30 +23,61 @@ namespace ckdb
 extern "C" {
 #endif
 
+#define ELEKTRA_NOTIFICATION_REGISTERFUNC_TYPEDEF(FUNC_TYPE_NAME, TYPE)                                                                    \
+	typedef int (*FUNC_TYPE_NAME) (Plugin * handle, Key * key, TYPE * variable);
+
+#define ELEKTRA_NOTIFICATION_TYPE_DEFINITION(TYPE, TYPE_NAME)                                                                              \
+	ELEKTRA_NOTIFICATION_REGISTER_SIGNATURE (TYPE, TYPE_NAME)                                                                          \
+	{                                                                                                                                  \
+		if (!kdb || !key || !variable)                                                                                             \
```
[the remark this answers is not in the document]
Comment (waht) [reply src/include/kdbnotificationinternal.h]: Yes, but I did not use it because this function is user facing.

File: src/plugins/internalnotification/macros/addType.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
+#ifndef TYPE
```
Comment (markus2330) [inline src/plugins/internalnotification/macros/addType.h]: Please all-lowercase filenames.

Please add some docu how to use the supermacro.

Will it work for @domhof (#1605) and to replace support.setfuncname in src/tools/gen/template/template.h?

File: src/plugins/internalnotification/macros/addType.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef TO_VALUE
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef CHECK_CONVERSION
+#define CHECK_CONVERSION 1
+#endif
+
+#define CONCAT(X, Y) CONCAT2 (X, Y)
```
Comment (markus2330) [inline src/plugins/internalnotification/macros/addType.h]: Is this needed as macro?

File: src/plugins/internalnotification/macros/addType.h [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	/* convert string to target type */
+	VALUE_TYPE value = TO_VALUE;
+	/* only update if conversion was successful */
+	if (CHECK_CONVERSION)
+	{
+		*(variable) = value;
+	}
+	else
+	{
+		ELEKTRA_LOG_WARNING ("conversion failed! string=%s, stopped=%c errno=%d", keyString (key), *end, errno);
```
Comment (markus2330) [inline src/plugins/internalnotification/macros/addType.h]: So the value is silently (unless the logger is enabled) not changed?

Not so ideal...

Comment (markus2330) [reply src/plugins/internalnotification/macros/addType.h]: > (e.g. the type plugin) to allow only valid values the key database.

Yes, that is the goal. But there is no type validation for C types.

> elektraNotificationSetConversionFailedCallback

It could be useful for applications to log and abort. But it is much better if such situations cannnot occur :-)

File: src/plugins/internalnotification/macros/addType.h (hunk 9 above)
Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: Yes. It concatenates two tokens. It was not necessary before since it was already encapsulated in a macro before expansion. I'm not sure how to describe, so here is an example :)

```C
// in supermacro.h
#define CONCAT(X, Y) CONCAT2 (X, Y)
#define CONCAT2(X, Y) X##Y
#define FUNC_NAME(TYPE_NAME) CONCAT (internalnotificationRegister, TYPE_NAME)
#define FUNC_NAME2(TYPE_NAME) internalnotificationRegister##TYPE_NAME
#define NESTED(TYPE_NAME) FUNC_NAME2 (TYPE_NAME)

FUNC_NAME (TYPE_NAME)
FUNC_NAME2 (TYPE_NAME)
NESTED (TYPE_NAME)

// in some.c
#define TYPE_NAME Int
#include "macros/supermacro.h"
// => gcc -E some.c
internalnotificationRegisterInt
internalnotificationRegisterTYPE_NAME
internalnotificationRegisterInt
```

Before the "supermacro", the `NESTED` macro was the macro that generated functions. A solution would be to add another macro (e.g. `#define FUNC_SIGNATURE(TYPE_NAME) void FUNC_NAME(TYPE_NAME) (void)` but the `CONCAT` macro is more versatile.

Comment (markus2330) [reply src/plugins/internalnotification/macros/addType.h]: Okay, then we should also add `ELEKTRA_CONCAT` next to `ELEKTRA_STRINGIFY`.

File: src/plugins/internalnotification/macros/addType.h (hunk 8 above)
Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: While my macros were inspired by his PR I think the only reusable part would be the conversion function. If needed we can extract it but details like error handling would need to be discussed. Theoretically we could have a lib dedicated to string to value and value to string conversions.

Comment (markus2330) [reply src/plugins/internalnotification/macros/addType.h]: Yes, let us discuss this.

Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: Ok! I'll think about error handling and reply not later than Thursday.

**Comment by markus2330** [comment]
Sorry, it seems like I created a merge conflict. Please rebase.

File: src/plugins/internalnotification/macros/addType.h (hunk 8 above)
Comment (markus2330) [reply src/plugins/internalnotification/macros/addType.h]: > undefined in case of an error.

If possible, the value should stay unchanged.

> What should be name of the conversion library?

If you can implement it as (super)macros no extra library is needed. Otherwise elektra-type might be suitable. We should also add `keySetLong`, `keyGetLong`, etc. functions.

Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: > If possible, the value should stay unchanged.

Ok its better to have additional variable declarations & logic in the macro instead of scattering it in the code base.

> If you can implement it as (super)macros no extra library is needed.

Of course! :) A naming macro as input would be required though to avoid name collisions.

> We should also add `keySetLong`, `keyGetLong`, etc. functions.

For such functions a seperate library would be better suited. Also I think this would broaden the scope of the PR too much. For now I'll better stick to the conversion macros.

Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: I added a supermacro called `type_create_to_value` for creating "to value" conversion functions in "include/macros". I decided to return the error and have a pointer for the variable because it improves readability:

```C
int variable = 0;
if (!myIntConversion (key, &variable))
{
    // error, variable is still 0
}
```

is better than
```C
int variable = 0;
int error = 0;
variable = myIntConversion (key, error)
if (error)
{
    // error, variable was changed to 0 because of return
}
```

Comment (markus2330) [reply src/plugins/internalnotification/macros/addType.h]: >  For such functions a seperate library would be better suited. Also I think this would broaden the scope of the PR too much. For now I'll better stick to the conversion macros.

Yes, of course! I am surprised that you put so much effort in supporting all types. I only mentioned keySetLong to make clear what elektra-type could be used for. But header-only without any lib is the better solution for this case.

Comment (waht) [reply src/plugins/internalnotification/macros/addType.h]: Ok, great! Then the PR should be completed :)

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
jenkins build gcc-configure-debian-debug please

**Comment by waht** [comment]
jenkins build mingw64 please

**Comment by waht** [comment]
The PR is ready for review. Is it safe to rebase now?

**Comment by markus2330** [comment]
Yes, release is done, so it is now a good time for rebase!

**Comment by waht** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by waht** [comment]
jenkins build multiconfig-gcc-stable please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by ingwinlu** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
@markus2330 I added the "reload" example and updated the tutorial, please review.
This PR should be merged before #1835 due to the kdbio-header reorganisation.

PS: Is the "build all" job supposed to be only "pr-merge" and "configure-debian-musl"?

**Comment by markus2330** [comment]
Yes, @ingwinlu nearly is finished with the migration to move all build jobs into one. Seems like musl is the last of the old ones.

File: src/include/kdbio/ev.h [hunk 11]
```diff
@@ -8,7 +8,7 @@
 #ifndef KDB_IOWRAPPER_EV_H_
 #define KDB_IOWRAPPER_EV_H_
 
-#include "kdbio.h"
+#include "../kdbio.h"
```
Comment (markus2330) [inline src/include/kdbio/ev.h]: Better to manipulate the include path and include kdbio/kdbio.h here.
I would even keep the kdbio.h top-level so that `<kdbio.h>` can be included.

File: doc/tutorials/notifications.md [hunk 12]
```diff
@@ -257,6 +257,116 @@ int main (void)
 }
 ```
 
+### How-To: Reload KDB when Elektra's configuration has changed
+
+This section shows how the notification feature can be used to reload an
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: is used to

File: doc/tutorials/notifications.md [hunk 13]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+application's KDB instance when Elektra's configuration has changed.
+This enables applications to apply changes to mount points or globally mounted
+plugins without restarting.
+
+**Step 1: Register for changes to Elektra's configuration**
+
+To achieve reloading on Elektra configuration changes we register for changes
+below the key `system/elektra` using
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: Does `/elektra` works too? It might be necessary once we have #1074

File: doc/tutorials/notifications.md [hunk 14]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+```
+
+**Step 2: Create a function for reloading KDB**
+
+Since our application needs to repeatedly initialize KDB on
+configuration changes we need to create a function which cleans
+up and reinitializates KDB.
+
+```C
+void initKdb (void)
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: `KDB ** kdb` as argument?

File: doc/tutorials/notifications.md [hunk 15]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+		// Cleanup notifications and close KDB
+		elektraNotificationClose (kdb);
+		kdbClose (kdb, parentKey);
+	}
+
+	kdb = kdbOpen (parentKey);
+	elektraIoSetBinding (kdb, binding);
+	elektraNotificationOpen (kdb);
+
+  // registration code from snippet before
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: indent?

File: doc/tutorials/notifications.md [hunk 16]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+This timer will then call the initialization function.
+
+First, we create the timer in the main loop setup of the application.
+
+```C
+// (global declaration)
+ElektraIoTimerOperation * reload;
+
+// main loop setup (e.g. main())
+reload = elektraIoNewTimerOperation (100, 0, initKdb, NULL);
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: Why 100?

File: doc/tutorials/notifications.md [hunk 17]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+void initKdb (void)
+{
+  // Stop reload task
+  elektraIoTimerSetEnabled (reload, 0);
+  elektraIoBindingUpdateTimer (reload);
+
+	if (kdb != NULL)
+	{
+		// Cleanup notifications and close KDB
+    elektraNotificationClose (kdb);
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: indent

File: doc/tutorials/notifications.md [hunk 18]
```diff
@@ -257,6 +257,116 @@ int main (void)
...
+		kdbClose (kdb, parentKey);
+	}
+
+  // ...
+}
+```
+
+By following these three steps any application can react to changes to Elektra's
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: "following" confusing here (usually refers to text afterwards). Better to use `By correct application of these ...`

File: examples/notificationAsync.c [hunk 19]
```diff
@@ -1,7 +1,17 @@
...
  *
- * @brief Implementation of notification functions as defined in kdbnotification.h
+ * @brief Example for notification library which repeatedly reads some keys and
+ * reacts to them
+ *
+ * Requires:
+ *   - io_uv binding
+ *   - Transport plugins (e.g. kdb global-mount dbus announce=once dbusrecv)
+ *
+ * Ideas for this example:
```
Comment (markus2330) [inline examples/notificationAsync.c]: Relevant keys for this example:

File: examples/notificationAsync.c [hunk 20]
```diff
@@ -100,7 +110,7 @@ int main (void)
 	int result = elektraNotificationOpen (kdb);
 	if (!result)
 	{
-		printf ("could init notification. aborting\n");
+		printf ("could not init notification. aborting\n");
```
Comment (markus2330) [inline examples/notificationAsync.c]: comma

File: examples/notificationPolling.c [hunk 21]
```diff
@@ -1,7 +1,13 @@
 /**
  * @file
  *
- * @brief Implementation of notification functions as defined in kdbnotification.h
+ * @brief Example for notification library which repeatedly reads some keys and
+ * reacts to them; see "example_notification_async" for an example without polling
+ *
+ * Ideas for this example:
```
Comment (markus2330) [inline examples/notificationPolling.c]: See above.

File: examples/notificationPolling.c [hunk 22]
```diff
@@ -94,7 +100,7 @@ int main (void)
 	int result = elektraNotificationOpen (kdb);
 	if (!result)
 	{
-		printf ("could init notification. aborting\n");
+		printf ("could not init notification. aborting\n");
```
Comment (markus2330) [inline examples/notificationPolling.c]: see above

File: examples/notificationReload.c [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+ * @file
+ *
+ * @brief Example for notification library which reloads KDB when Elektra's
+ * configuration (e.g. mount points or global plugins) has changed.
+ *
+ * Requires:
+ *   - io_glib binding
+ *   - Transport plugins (e.g. kdb global-mount zeromqsend zeromqrecv && kdb run-hub-zeromq)
+ *
+ * Ideas for this example:
```
Comment (markus2330) [inline examples/notificationReload.c]: see above

File: examples/notificationReload.c [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+		printf ("could not open KDB. aborting\n");
+		exit (-1);
+	}
+
+	elektraIoSetBinding (kdb, binding);
+
+	int result = elektraNotificationOpen (kdb);
+	if (!result)
+	{
+		printf ("could not init notification. aborting\n");
```
Comment (markus2330) [inline examples/notificationReload.c]: see above

File: examples/notificationReload.c [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+		elektraNotificationClose (kdb);
+		kdbClose (kdb, parentKey);
+		didReload = 1;
+	}
+
+	kdb = kdbOpen (parentKey);
+	if (kdb == NULL)
+	{
+		printf ("could not open KDB. aborting\n");
+		exit (-1);
```
Comment (markus2330) [inline examples/notificationReload.c]: This would return "status & 0377" to the parent. Why not using exit (1)?

File: doc/news/_preparation_next_release.md [hunk 26]
```diff
@@ -223,6 +232,20 @@ Thanks to Michael Zronek and Vanessa Kos.
 
 [#1887]: https://github.com/ElektraInitiative/libelektra/issues/1887
 
+## Compatibility
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: I think the "Compatibility" above happened at some wrong merge. @sanssecours any idea how this happened?

File: examples/CMakeLists.txt [hunk 27]
```diff
@@ -28,3 +31,74 @@ target_link_elektra (kdbintro elektra-kdb)
...
+
+	set (SRC_FILES notificationPolling.c)
+	set (SOURCES ${SRC_FILES} ${HDR_FILES})
+
+	add_executable (${EXAMPLE} ${SOURCES})
+	add_dependencies (${EXAMPLE} kdberrors_generated)
+
+	target_link_elektra (${EXAMPLE} elektra-kdb elektra-notification)
+
+	# TODO resolve https://github.com/ElektraInitiative/libelektra/issues/2007
```
Comment (markus2330) [inline examples/CMakeLists.txt]: please use URL issues.libelektra.org/2007

File: examples/notificationAsync.c [hunk 28]
```diff
@@ -115,13 +135,22 @@ int main (void)
...
 	printf ("- Set \"%s\" to red, blue or green to change the text color\n", keyName (callbackKeyToWatch));
 	printf ("- Set \"%s\" to any integer value\n", keyName (intKeyToWatch));
 	printf ("Send SIGINT (Ctl+C) to exit.\n\n");
 
+	// Get configuration
+	kdbGet (kdb, config, key);
+	printVariable (timer); // "value" was automatically updated
+
+	// This allows us to wake the loop from our signal handler
+#ifdef HAVE_LIBUV1
```
Comment (markus2330) [inline examples/notificationAsync.c]: @ingwinlu do we have libuv1 installed somewhere?

File: src/include/kdbnotification.h [hunk 29]
```diff
@@ -101,37 +104,95 @@ int elektraNotificationOpen (KDB * kdb);
...
+ * @param  variable variable
  *
  * @retval 1 on success
  * @retval 0 on failure
+ * @{
+ */
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (int, Int)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (unsigned int, UnsignedInt)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (long, Long)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (unsigned long, UnsignedLong)
```
Comment (markus2330) [inline src/include/kdbnotification.h]: why not long long?

File: src/include/macros/type_create_to_value.h [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+ * @param  VALUE_TYPE       optional, defaults to TYPE. Ideally a larger type assigned to variable `value` for
+ *                          checking the range before the variable is updated
+ * @param  TO_VALUE         expression for converting `string` (variable containing the key value) to VALUE_TYPE
+ * @param  CHECK_CONVERSION optional, defaults to true. A boolean expression. Allows to check the range after
+ *                          conversion. Use ELEKTRA_TYPE_CHECK_CONVERSION to check if a conversion using
+ *                          strto*()-functions was successful and ELEKTRA_TYPE_CHECK_CONVERSION_RANGE (RANGE)
+ *                          to check additionally for a specified range.
+ * @param DISABLE_UNDEF_PARAMETERS define to disable undefining of parameters after the macro. Use if parameters
+ *                                 are used within another supermacro.
+ */
```
Comment (markus2330) [inline src/include/macros/type_create_to_value.h]: license BSD missing

File: doc/news/_preparation_next_release.md (hunk 26 above)
Comment (sanssecours) [reply doc/news/_preparation_next_release.md]: I think the duplicated header was added in commit 1b2d0d62. Anyway, pull request [#2066](https://github.com/ElektraInitiative/libelektra/pull/2066/files#diff-47ed151a0df6c926f5b1320d140a5b34L227) should fix the problem.

File: doc/tutorials/notifications.md (hunk 13 above)
Comment (waht) [reply doc/tutorials/notifications.md]: Yes, `/elektra` works too. I'll update the example.

File: doc/tutorials/notifications.md (hunk 16 above)
Comment (waht) [reply doc/tutorials/notifications.md]: The timer interval is 100 milliseconds; I'll add a comment.

File: examples/notificationReload.c (hunk 25 above)
Comment (waht) [reply examples/notificationReload.c]: No reason really, except the other examples used "return -1"; I'll use `exit (1)`

File: examples/notificationAsync.c (hunk 28 above)
Comment (waht) [reply examples/notificationAsync.c]: libuv 1.x was included in the debian docker images, but afaik they are not used by the build server.
On Travis we only have libuv 0.x available. This PR will add 1.x to the build server.

File: src/include/kdbnotification.h (hunk 29 above)
Comment (waht) [reply src/include/kdbnotification.h]: No problem, I'll add it.

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by waht** [comment]
@markus2330 I fixed some unsigned conversion issues, the PR is ready for a re-review :)

File: src/plugins/internalnotification/README.md [hunk 31]
```diff
@@ -19,47 +19,9 @@ Application developers should use the
...
-When the key changes the callback is called with the new key.
-
-*Additional Parameters*
-
-- *callback* Callback function with the signature
-    `void (*ElektraNotificationChangeCallback) (Key * key)`.
+This plugin exports various functions starting with "register*" below
+`system/elektra/modules/internalnotification/exports/`.
+These functions should not be used directly.
+Instead the notification API should be used.
```
Comment (markus2330) [inline src/plugins/internalnotification/README.md]: Link to notification API.

File: src/libs/notification/example/README.md [hunk 32]
```diff
@@ -1,28 +0,0 @@
-# Example Applications for Notifications
```
Comment (markus2330) [inline src/libs/notification/example/README.md]: Why is this README removed?

Btw. the  src/libs/notification/example is quite hidden, maybe you can add links to it also at other places.

File: src/libs/invoke/invoke.c [hunk 33]
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
...
+	{
+		return;
+	}
+	elektraDeferredCallsExecute (handle->plugin, list);
+}
+
+/**
+ * Call a deferable function on a plugin handle.
+ * If the function is exported by the plugin it is directly invoked,
+ * if the plugin supports deferring calls, the call is deferred.
```
Comment (markus2330) [inline src/libs/invoke/invoke.c]: What if it is exported and also supports deferring calls?

File: src/libs/invoke/invoke.c [hunk 34]
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
...
+ *
+ * The parameters key set can be freed afterwards.
```
Comment (markus2330) [inline src/libs/invoke/invoke.c]: Write this directly to the param description.

File: src/libs/invoke/invoke.c [hunk 35]
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
...
+ *
+ * @param  handle                    invoke handle
+ * @param  elektraPluginFunctionName function name
+ * @param  parameters                parameter key set
+ * @retval 0 on success
+ * @retval 1 when function not exported and deferring is unsupported by plugin
```
Comment (markus2330) [inline src/libs/invoke/invoke.c]: should be negative (like other errors). How can the user know which situation happened?

File: src/libs/README.md [hunk 36]
```diff
@@ -82,3 +82,15 @@ data structures.
...
+
+    libelektra-io.so
+
+**[io](io/)** provides the common [I/O binding API](https://doc.libelektra.org/api/current/html/group__kdbio.html).
+
+### Notification
+
+    libelektra-notification.so
+
+**[notification](notification/)** provides the [notification API](https://doc.libelektra.org/api/current/html/group__kdbnotification.html).
```
Comment (markus2330) [inline src/libs/README.md]: Maybe add link to examples also here.

File: src/libs/README.md [hunk 37]
```diff
@@ -82,3 +82,15 @@ data structures.
     libelektra-invoke.so
 
 **[libinvoke](invoke/)** provides a simple API allowing us to call functions exported by plugins.
+
+### IO
+
+    libelektra-io.so
+
+**[io](io/)** provides the common [I/O binding API](https://doc.libelektra.org/api/current/html/group__kdbio.html).
```
Comment (markus2330) [inline src/libs/README.md]: a very short description would be nice. It is unclear what "I/O" is in this context.

File: examples/notificationReload.c [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+#include <stdio.h>  // printf() & co
+#include <stdlib.h> // exit()
+
+#define TWO_SECONDS 2000
+#define RELOAD_INTERVAL 100
+
+GMainLoop * loop;
+
+KDB * kdb;
+Key * parentKey;
```
Comment (markus2330) [inline examples/notificationReload.c]: are all of them needed globally? Otherwise the example is great!

File: doc/tutorials/notifications.md [hunk 39]
```diff
@@ -53,125 +47,163 @@ notifications but not whether notifications are sent or which transport is used.
...
-#include <elektra/kdbio_uv.h>
-#include <elektra/kdbnotification.h>
+#include <elektra/kdbio/uv.h>
 
 #include <uv.h>
 
 void main (void)
 {
-	KDB* repo;
+  KDB* repo;
```
Comment (markus2330) [inline doc/tutorials/notifications.md]: Any specific reason why the change to spaces here?

Comment (waht) [reply doc/tutorials/notifications.md]: No, no reason. Might be the editor messing with identation or me forgetting about it when pasting code.

File: src/libs/notification/example/README.md (hunk 32 above)
Comment (waht) [reply src/libs/notification/example/README.md]: I forgot to delete the folder - I moved them to /examples. 

About the README: The existing examples didn't have one so I removed it. On second thought I should add instructions about expected plugins to the examples (i.e. do a `printf()`).

File: examples/notificationReload.c (hunk 38 above)
Comment (waht) [reply examples/notificationReload.c]: Thank you! No they can also be moved to a struct which is passed as user data - I updated the example to demo usage of the required functions (`elektraIo[Binding|Fd|Timer|Idle]GetData()`).

File: src/libs/invoke/invoke.c (hunk 33 above)
Comment (waht) [reply src/libs/invoke/invoke.c]: The code currently favors direct calls and does not defer it. I gave it some thought and it would be useful to do both when possible. 
If, for example, the list plugin used I/O bindings and exposed a "setIoBinding" function then it would be useful to still have the call deferred to its encapsulated plugins.

File: src/libs/invoke/invoke.c (hunk 35 above)
Comment (waht) [reply src/libs/invoke/invoke.c]: I'm not sure if the user needs to know what happened. I think it is only interesting to know whether the call failed (cannot deferr and cannot call directly) or whether there was success (either deferred or directly called [or both]).

File: src/libs/notification/example/README.md (hunk 32 above)
Comment (waht) [reply src/libs/notification/example/README.md]: Here is an updated example: https://github.com/ElektraInitiative/libelektra/pull/1850/files#diff-98f467ea2a20eaed5714de95cc0d4142R193

File: examples/notificationReload.c (hunk 38 above)
Comment (waht) [reply examples/notificationReload.c]: Please check out the [updated example](https://github.com/ElektraInitiative/libelektra/pull/1850/files#diff-98f467ea2a20eaed5714de95cc0d4142).

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
Looks much better now, is it ready to merge?

**Comment by markus2330** [comment]
@tom-wa would be great to have this for the upcoming release which hopefully happens today.

**Comment by markus2330** [comment]
Sorry, I wanted to ping @waht here.

**Comment by waht** [comment]
Sure! I just added a commit improving the connection timeout resolution.

**Comment by markus2330** [comment]
Thank you, finally it is done :smile:
